src/cmd-tag

```diff
@@ -0,0 +1,177 @@
+#
+#   $ coreos-assembler tag update --tag smoketested --build 47.152
+#   $ coreos-assembler tag delete --tag beta
+#   $ coreos-assembler tag list
+#
+import argparse
+import json
+import os
+import sys
+
+sys.path.insert(0, '/usr/lib/coreos-assembler')
+from cmdlib import fatal, rfc3339_time, write_json
+
+
+def main():
+    """Main entry point"""
+
+    args = parse_args()
+    args.func(args)
+
+
+def parse_args():
+    """Parse args and dispatch"""
+
+    parser = argparse.ArgumentParser()
+    subparsers = parser.add_subparsers(dest="cmd", title="Tag actions")
+    subparsers.required = True
+
+    delete = subparsers.add_parser("delete",
+                                   help="Delete a tag from the metadata")
+    delete.add_argument("--tag", help="Tag to delete from metadata",
+                        required=True)
+    delete.set_defaults(func=cmd_delete)
+
+    list_tags = subparsers.add_parser("list", help="List available tags in "
+                                                   "the metadata")
+    list_tags.set_defaults(func=cmd_list)
+
+    update = subparsers.add_parser("update", help="Update existing tag or "
+                                                  "create a new tag")
+    update.add_argument("--tag", help="Tag to be updated", required=True)
+    update.add_argument("--build", help="Build to update tag with",
+                        required=True)
+    update.add_argument("--force", help="Force the update of a tag",
+                        action="store_true")
+    update.set_defaults(func=cmd_update)
+
+    return parser.parse_args()
+
+
+def init_build_data():
+    """Initialize build metadata"""
+
+    builds_json = "builds/builds.json"
+
+    if os.path.isfile(builds_json):
+        with open(builds_json) as json_file:
+            build_data = json.load(json_file)
+    else:
+        print("INFO: Did not find existing builds.json; starting with empty "
+              "build list and tags list")
+        build_data = {"builds": [], "tags": []}
+
+    return build_data
+
+
+def finalize_json(build_data):
+    """Wrapper around cmdlib.write_json()"""
+
+    builds_json = "builds/builds.json"
+
+    build_data['timestamp'] = rfc3339_time()
+    write_json(builds_json, build_data)
+
+
+def available_builds(build_data):
+    """Returns the availble build IDs from build metadata"""
+
+    return build_data["builds"]
+
+
+def available_tags(build_data):
+    """Returns the available tag names from build metadata"""
+
+    return [t.get("name") for t in build_data.get("tags", {})]
+
+
+def cmd_list(args):
+    """List available tags in build metadata"""
+
+    build_data = init_build_data()
+    avail_tags = available_tags(build_data)
+    if not avail_tags:
+        print("No tags found")
+        return
+
+    for tag in build_data["tags"]:
+        print(json.dumps({"name": tag.get("name"),
+                          "created": tag.get("created"),
+                          "target": tag.get("target")}))
```

Comment:
My original suggestion was to put it all in one JSON output like @wking mentioned so that it could be `jq` parse-able. If the goal is to have more human readable output natively though I think the original solution is fine.

Reply:
Apologies, I misunderstood the original suggestion @arithx.

I think tagging will likely be something that higher-order build systems are doing (i.e. not the user doing builds locally on their host) and those build systems would have prior knowledge of the tags being used.  I added the `list` verb with humans in mind (i.e. debugging tags?), so I'm in favor of human-readable output.

However, I'm willing to switch to something `jq` parse-able if others find it useful.

Reply:
I'm fine with whatever output, but personally I'd be much more likely to use `jq` directly on `builds.json` for this and skip Python altogether.  So don't worry about supporting my tag-listing needs ;).

Reply:
Yeah, fair to switch to the human readable output and if we decide we want to add it into the python to do some extra modifications later we can just add a flag to output JSON.

Reply:
Thanks for the feedback folks!  Will leave it as is for now.  :+1: